monitoring/ping.go

```diff
@@ -0,0 +1,263 @@
+		if self.Addr.String() == node.Addr.String() {
+			continue
+		}
+
+		rttNanoSec, err := calculateRTT(client, self, node)
+		if err != nil {
+			return err
+		}
+
+		err = c.storePingInHDR(rttNanoSec, node)
+		if err != nil {
+			return err
+		}
+
+		roundtripLatencyInterface, _ := c.roundtripLatency.Get(node.Name)
+		roundtripLatency, ok := roundtripLatencyInterface.(*hdrhistogram.Histogram)
+		if !ok {
+			return trace.Errorf("couldn't parse roundtripLatency as HDRHistogram on %s", c.serfMemberName)
+		}
+		log.Debugf("%s <-ping-> %s = %dns [latest]", self.Name, node.Name, rttNanoSec)
+		log.Debugf("%s <-ping-> %s = %dns [%.2f percentile]",
+			self.Name, node.Name,
+			roundtripLatency.ValueAtQuantile(pingRoundtripQuantile),
+			pingRoundtripQuantile)
+
+		pingRoundtripPercentile := roundtripLatency.ValueAtQuantile(pingRoundtripQuantile)
+		if pingRoundtripPercentile >= pingRoundtripThreshold.Nanoseconds() {
+			log.Warningf("%s <-ping-> %s = slow ping RoundTrip detected. Value %dns over threshold %s (%dns)",
+				self.Name, node.Name, pingRoundtripPercentile,
+				pingRoundtripThreshold.String(), pingRoundtripThreshold.Nanoseconds())
+		} else {
+			log.Debugf("%s <-ping-> %s = ping RoundTrip okay. Value %dns within threshold %s (%dns)",
+				self.Name, node.Name, pingRoundtripPercentile,
+				pingRoundtripThreshold.String(), pingRoundtripThreshold.Nanoseconds())
+		}
+	}
+
+	return nil
+}
+
+// storePingInHDR is used to store ping RoundTrip values in HDR Histograms in memory
+func (c *pingChecker) storePingInHDR(pingroundtripLatency int64, node serf.Member) error {
+	s, exists := c.roundtripLatency.Get(node.Name)
+	if !exists {
+		c.roundtripLatency.Set(node.Name,
+			hdrhistogram.New(pingRoundtripMinimum.Nanoseconds(),
+				pingRoundtripMaximum.Nanoseconds(),
+				pingRoundtripSignificativeFigures),
+			statsTTLPeriod)
+		s, _ = c.roundtripLatency.Get(node.Name)
+	}
+
+	nodeLatencies, ok := s.(*hdrhistogram.Histogram)
+	if !ok {
+		return trace.BadParameter("couldn't parse roundtripLatency as HDRHistogram on %s", c.serfMemberName)
+	}
+
+	tmpSnapshot := nodeLatencies.Export()
+	if len(tmpSnapshot.Counts) >= slidingWindowSize {
+		// pop element at index 0 (oldest)
+		nodeLatencies.Reset()
+
+		lowerLimit := len(tmpSnapshot.Counts) - slidingWindowSize
+		if lowerLimit < 0 {
+			lowerLimit = 0
+		}
+		// using len(tmpSnapshot.Counts)-1 as the upper limit as another value
+		// will be added a few lines below
+		for i := lowerLimit; i < len(tmpSnapshot.Counts)-1; i++ {
+			nodeLatencies.RecordValue(tmpSnapshot.Counts[i])
+		}
+	}
+
+	err := nodeLatencies.RecordValue(pingroundtripLatency)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("%d recorded ping RoundTrip values for node %s",
+		nodeLatencies.TotalCount(), node.Name)
+
+	return nil
+}
+
+// calculateRTT calculates the RoundTrip time between two Serf Cluster members
+func calculateRTT(serfClient *serf.RPCClient, self serf.Member, node serf.Member) (int64, error) {
```

Comment:
nit: you can combine arguments of the same type:
```go
func calculateRTT(serfClient *serf.RPCClient, self, node serf.Member) (rttNanos int64, err error) {
```
Also note that it's recommended to name the returns for readability.